synapse/federation/sender/per_destination_queue.py

```diff
@@ -542,22 +550,53 @@ async def _catch_up_transmission_loop(self) -> None:
                     self._destination, last_successful_stream_ordering
                 )
 
-    def _get_rr_edus(self, force_flush: bool) -> Iterable[Edu]:
+    def _get_rr_edus(self, force_flush: bool, limit: int) -> Iterable[Edu]:
         if not self._pending_rrs:
             return
         if not force_flush and not self._rrs_pending_flush:
             # not yet time for this lot
             return
 
-        edu = Edu(
-            origin=self._server_name,
-            destination=self._destination,
-            edu_type=EduTypes.RECEIPT,
-            content=self._pending_rrs,
-        )
+        # Build the EDUs needed to send these receipts. This is a bit complicated
+        # since we can share one for each unique (room, receipt type, user), but
+        # need additional ones for different threads. The result is that we will
+        # send N EDUs where N is the number of unique threads across rooms.
+        #
+        # This could be more efficient by bundling users who have sent receipts
+        # for different threads.
+        generated_edus = 0
+        while self._pending_rrs and generated_edus < limit:
+            # The next EDU's content.
+            content: JsonDict = {}
+
+            # Iterate each room's receipt types and threads, adding it to the content.
+            for room_id in list(self._pending_rrs.keys()):
+                for receipt_type in list(self._pending_rrs[room_id].keys()):
+                    thread_ids = self._pending_rrs[room_id][receipt_type]
+                    # The thread ID itself doesn't matter at this point.
+                    content.setdefault(room_id, {})[
+                        receipt_type
+                    ] = thread_ids.popitem()[1]
+
+                    # If there are no threads left in this room / receipt type.
+                    # Clear it out.
+                    if not thread_ids:
+                        del self._pending_rrs[room_id][receipt_type]
+
+                # Again, clear out any blank rooms.
+                if not self._pending_rrs[room_id]:
+                    del self._pending_rrs[room_id]
+
+            yield Edu(
+                origin=self._server_name,
+                destination=self._destination,
+                edu_type=EduTypes.RECEIPT,
+                content=content,
+            )
+            generated_edus += 1
+
         self._pending_rrs = {}
```

Comment:
This is clearly wrong now as we might not have processed all read receipts. Will fix.